### ISSUE TITLE
Problem with custom loss

### ISSUE THREAD

**Issue by piccolbo** [wwoods/keras_pickle_wrapper#1]
I happen to have written a very similar class, but it doesn't work with custom loss functions because of keras#5916

I suspect your approach is also affected. I may be able to provide a test case later on, but I wanted to quickly put in a warning that it may be a problem. Thanks

**Comment by wwoods** [comment]
It has worked for me in the past... the key is registering the loss function with `KerasPickleWrapper.register()`.  I'll look more if you can show me it doesn't work when registered, I'll look into it more.

**Comment by piccolbo** [comment]
Good suggestion

**Comment by piccolbo** [comment]
Wait a second, do you need to call that at u pickle time?

**Comment by piccolbo** [comment]
I mean unpickle

**Comment by wwoods** [comment]
It's part of setup.  Whether you are pickling or unpickling, you must (before either) declare your custom classes and register them.  It's just the way Keras' name resolution works.  Typically, it's best to register the custom class immediately below the declaration, since it's all Keras-specific anyway.

**Comment by piccolbo** [comment]
Got it. My problem is that spark is doing the unpickling before any user
code. I have an ugly workaround but not a solution. Thanks for
clarification.

**Comment by wwoods** [comment]
Just for the record... Your user code MUST be running first.  Unpickling will never work if a custom class isn't defined.  However, the way Python resolves those names is more sane than Keras' approach, and thus you do not need to register the classes.

If spark is unpickling a custom function of yours, then your loss function must be defined before the unpickle.  That would be the place to register.

Another workaround, if spark truly only supports pickling built-in Python objects and not user-custom objects (which would be weird, but I haven't used spark much):

This keras-pickle-wrapper by default unpickles unloaded.  That is, it is OK to unpickle a KerasPickleWrapper object, then call `KerasPickleWrapper.register()` for the appropriate custom objects, then call (load) the KerasPickleWrapper.  It's just the way I wrote the wrapper - lazy unpickling of the Keras object.

**Comment by piccolbo** [comment]
For the record, I just have a custom loss, no custom class.

**Comment by wwoods** [comment]
That's still a function, and should be defined similarly to a class.  You
should be able to register it immediately after the definition; pickled
functions just refer to their fully-qualified name:
https://docs.python.org/3.6/library/pickle.html#what-can-be-pickled-and-unpickled
.
